sql/sql_show.cc

```diff
@@ -4653,6 +4653,129 @@ fill_schema_table_by_open(THD *thd, MEM_ROOT *mem_root,
+  @param[in]      tables                   I_S table
+  @param[in]      cond                     'WHERE' condition
+  @param[in]      table_name               table name
+
+  @return         Operation status
+  @retval         0                  success
+  @retval         1                  error
+*/
+
+static int fill_global_temporary_tables(THD *thd, TABLE_LIST *tables, Item *cond)
+{
+  DBUG_ENTER("fill_global_temporary_tables");
+  mysql_mutex_lock(&LOCK_thread_count);
+
+  CHARSET_INFO *cs= system_charset_info;
+  THD *thd1;
+  I_List_iterator<THD> it(threads);
+
+#ifndef NO_EMBEDDED_ACCESS_CHECKS
+  Security_context *sctx= thd->security_ctx;
+  uint db_access;
+#endif
+
+  while ((thd1= it++))
+  {
+
+#ifndef DBUG_OFF
+    const char *tmp_proc_info= thd1->proc_info;
+    if (tmp_proc_info && !strncmp(tmp_proc_info, STRING_WITH_LEN("debug sync point: before_open_in_get_all_tables")))
+    {
+      DEBUG_SYNC(thd1, "fill_global_temporary_tables_thd_item_at_tables_debug_sync");
+    }
+#endif
+
+    if (thd1->temporary_tables!= 0)
+    {
+      mysql_mutex_lock(&thd1->LOCK_temporary_tables);
+      All_tmp_tables_list *tl= thd1->temporary_tables;
+      TMP_TABLE_SHARE *tmp= (*tl).front();
+
+      while (tmp!= 0)
+      {
+#ifndef NO_EMBEDDED_ACCESS_CHECKS
+        if (test_all_bits(sctx->master_access, DB_ACLS))
+          db_access= DB_ACLS;
+        else
+          db_access= (acl_get(sctx->host, sctx->ip,
+                              sctx->priv_user, (*tmp).db.str, 0)
+                      | sctx->master_access);
+        if (!(db_access & DB_ACLS) && check_grant_db(thd1, (*tmp).db.str))
+        {
+          //no access for temp tables within this db for user
+          continue;
+        }
+#endif
+
+        DEBUG_SYNC(thd1, "fill_global_temporary_tables_before_storing_rec");
+        restore_record(tables->table, s->default_values);
+        // session_id
+        tables->table->field[0]->store((longlong) thd1->thread_id, TRUE);
+        // table_schema
+        tables->table->field[1]->store((*tmp).db.str, (*tmp).db.length, cs);
+        // table_name
+        tables->table->field[2]->store((*tmp).table_name.str, (*tmp).table_name.length, cs);
+        // engine
+        TABLE *current_table= tmp->all_tmp_tables.front();
+
+        handler *handle= current_table->file;
+        // Assume that invoking handler::table_type() on a shared handler is safe
+        const char *engine_type= (char*)(handle ? handle->table_type(): "UNKNOWN");
+        tables->table->field[3]->store(engine_type,strlen(engine_type), cs);
+        // name
+        const char *path= strstr((*tmp).path.str, "#sql");
+        int len= (*tmp).path.length-(path-(*tmp).path.str);
+        tables->table->field[4]->store(path, len, cs);
+        // File stats
+        handler *file= current_table->db_stat ? current_table->file : 0;
+        if (file)
+        {
+          file= file->clone((*current_table).s->normalized_path.str, thd->mem_root);
```

Comment:
Why are you cloning file here? Could you just use the stats. Or copy the stats like a structure if there's an atomic risk of it going away.

Reply:
It looks like it's taken or at least very similar to Percona Server implementation, hence, I can tell why  it's being cloned - due to file->info call below. If you don't clone, calling info is not safe. If you don't clone and don' call info, you don't always have useful information to return. git blame on our repo here should have bug references and discussion of this

Reply:
ack. thanks for the references.

Reply:
There is no reason to protect against file->info().
file->info() is updating a local area of the table.  It will not affect other handlers using the same file
If you are trying to look at a temporary file that is in use by someone else, you should not call file->info() at all
but instead use the current status.

Generally I don't think it's a good idea to be able to look at temporary files used by other users as there is no protection that someone can delete them while you are using them.  Some engines probably does not allow you to do a clone() of all types of temporary tables as the assumption for these is that there can only be one user for them.
For example, internal temporary tables (used for temporary results) would not allow clone().